library/std/src/thread/scoped.rs

```diff
@@ -0,0 +1,301 @@
+/// #![feature(scoped_threads)]
+/// use std::thread;
+///
+/// let mut a = vec![1, 2, 3];
+/// let mut x = 0;
+///
+/// thread::scope(|s| {
+///     s.spawn(|_| {
+///         println!("hello from the first scoped thread");
+///         // We can borrow `a` here.
+///         dbg!(&a);
+///     });
+///     s.spawn(|_| {
+///         println!("hello from the second scoped thread");
+///         // We can even mutably borrow `x` here,
+///         // because no other threads are using it.
+///         x += a[0] + a[2];
+///     });
+///     println!("hello from the main thread");
+/// });
+///
+/// // After the scope, we can modify and access our variables again:
+/// a.push(4);
+/// assert_eq!(x, a.len());
+/// ```
+#[track_caller]
+pub fn scope<'env, F, T>(f: F) -> T
+where
+    F: FnOnce(&Scope<'env>) -> T,
+{
+    let scope = Scope {
+        data: ScopeData {
+            n_running_threads: AtomicUsize::new(0),
+            main_thread: current(),
+            a_thread_panicked: AtomicBool::new(false),
+        },
+        env: PhantomData,
+    };
+
+    // Run `f`, but catch panics so we can make sure to wait for all the threads to join.
+    let result = catch_unwind(AssertUnwindSafe(|| f(&scope)));
+
+    // Wait until all the threads are finished.
+    while scope.data.n_running_threads.load(Ordering::Acquire) != 0 {
+        park();
+    }
+
+    // Throw any panic from `f`, or the return value of `f` if no thread panicked.
+    match result {
+        Err(e) => resume_unwind(e),
+        Ok(_) if scope.data.a_thread_panicked.load(Ordering::Relaxed) => {
+            panic!("a thread panicked")
+        }
+        Ok(result) => result,
+    }
+}
+
+impl<'env> Scope<'env> {
+    /// Spawns a new thread within a scope, returning a [`ScopedJoinHandle`] for it.
+    ///
+    /// Unlike non-scoped threads, threads spawned with this function may
+    /// borrow non-`'static` data from the outside the scope. See [`scope`] for
+    /// details.
+    ///
+    /// The join handle provides a [`join`] method that can be used to join the spawned
+    /// thread. If the spawned thread panics, [`join`] will return an [`Err`] containing
+    /// the panic payload.
+    ///
+    /// If the join handle is dropped, the spawned thread will implicitly joined at the
+    /// end of the scope. In that case, if the spawned thread panics, [`scope`] will
+    /// panic after all threads are joined.
+    ///
+    /// This call will create a thread using default parameters of [`Builder`].
+    /// If you want to specify the stack size or the name of the thread, use
+    /// [`Builder::spawn_scoped`] instead.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the OS fails to create a thread; use [`Builder::spawn_scoped`]
+    /// to recover from such errors.
+    ///
+    /// [`join`]: ScopedJoinHandle::join
+    pub fn spawn<'scope, F, T>(&'scope self, f: F) -> ScopedJoinHandle<'scope, T>
+    where
+        F: FnOnce(&Scope<'env>) -> T + Send + 'env,
```

Comment:
It's a pity that this can't really be `&'scope Scope<'env>` (I mean, it can, but it will, in practice, contradict the `+ 'env` requirement).

  - This whole design would thus be improved if non-explicitly-`move` closure were able to capture an `impl Copy` by move (and thus copy) rather than by ref, at least for some types that could opt into that behavior (cc @nikomatsakis, since you've put a lot of thought into improving closure captures, and cc @chorman0773). I posted lengthily about this on one of the original scoped-thread-API issues, but repeating the main point here as well to keep raising awareness 🙂

Reply:
My $.02 on this comes from an abi perspective, rather than a language perpsective, but seeing that I've been summoned , I may as well share it.
It may not necessarily be a good idea to capture *all* `Copy` types by move implicitly. A `[u32;1024]` may be trivial to copy, but it's certainly not cheap to copy or to store. 
In the preliminary abi spec used for [lccc](https://github.com/LightningCreations/lccc)'s rust frontend, I've prescribed that small (up to pointer-sized) values of a `Copy` type get stored in the closure directly if the address is unused.
> A capture of a type that is at most the same size of a pointer, which implements Copy and which is neither borrowed nor mutated by the body of the closure is captured by-copy for the purposes of ABI, even if the closure would otherwise capture it by-reference.

(Hilareously, this used to apply *all* Copy types, before that defect was pointed out to me).

It may or may not be a good idea to apply this rule as a language rule, mostly because it can be fairly interesting what "neither borrowed nor mutated" means (passing an `i32` capture to `println!`, for example, would exclude the capture from this rule in the abi, since the macro internally takes the address of each argument, despite not actually caring), and it would be a breaking change to extend it in general to all `Copy` captures regardless of whether the address is used (you could, for example, have `println!("{:p}",&i);` where `i` is an `i32` capture).

Reply:
Yeah, that's why I mentioned the opt-in part: I don't think there could be an always-applicable case; an alternative to annotating a `Copy` type its preference in being capture by move rather than by ref would be having the option to pin-point the captures of a closure:

```rust
scope.spawn(move(scope) || { // <- capture it if spawning nested threads
    scope.spawn(|| { … }); // otherwise unneeded
})
```

For the scoped threads API it would thus be less pretty than having that `move(scope)` happen implicitly, but it would be a more generally-applicable and visible alternative.

Reply:
Note that this PR only adds this as unstable, meaning we can still change/improve its interface before stabilizing it, if something better becomes possible. Let's make sure to discuss this on the tracking issue once the RFC is merged.

Reply:
I've added this as an unresolved question to the tracking issue: https://github.com/rust-lang/rust/issues/93203